File: app/src/main/java/pl/szczodrzynski/edziennik/datamodels/Grade.java
package pl.szczodrzynski.edziennik.datamodels;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.Ignore;
import androidx.room.Index;

@Entity(tableName = "grades",
        primaryKeys = {"profileId", "gradeId"},
        indices = {@Index(value = {"profileId"})})
public class Grade {
    public int profileId;

    @ColumnInfo(name = "gradeId")
    public long id;

    @ColumnInfo(name = "gradeCategory")
    public String category;
    @ColumnInfo(name = "gradeColor")
    public int color;
    @ColumnInfo(name = "gradeDescription")
    public String description;
    @ColumnInfo(name = "gradeComment")
    public String comment;
    @ColumnInfo(name = "gradeName")
    public String name;
    @ColumnInfo(name = "gradeValue")
    public float value;
    @ColumnInfo(name = "gradeValueMax")
    public float valueMax;
    @ColumnInfo(name = "gradeWeight")
    public float weight;
    @ColumnInfo(name = "gradeSemester")
    public int semester;
    @ColumnInfo(name = "gradeClassAverage")
    public float classAverage = -1;
    public static final int TYPE_NORMAL = 0;
    public static final int TYPE_SEMESTER1_PROPOSED = 1;
    public static final int TYPE_SEMESTER1_FINAL = 2;
    public static final int TYPE_SEMESTER2_PROPOSED = 3;
    public static final int TYPE_SEMESTER2_FINAL = 4;
    public static final int TYPE_YEAR_PROPOSED = 5;
    public static final int TYPE_YEAR_FINAL = 6;
    public static final int TYPE_POINT = 10;
    public static final int TYPE_BEHAVIOUR = 20;
    public static final int TYPE_DESCRIPTIVE = 30;
    public static final int TYPE_TEXT = 40;
    @ColumnInfo(name = "gradeType")
    public int type = TYPE_NORMAL;
    @ColumnInfo(name = "gradePointGrade")
    public boolean pointGrade = false;

    /**
     * Applies for historical grades. It's the new/replacement grade's ID.
     */
    @ColumnInfo(name = "gradeParentId")
    public long parentId = -1;

    /**
     * Applies for current grades. If the grade was worse and this is the improved one.
     */
    @ColumnInfo(name = "gradeIsImprovement")
    public boolean isImprovement = false;

    public long teacherId;
    public long subjectId;

    @Ignore
    public Grade() {}

    public Grade(int profileId, long id, String category, int color, String description, String name, float value, float weight, int semester, long teacherId, long subjectId) {
        this.profileId = profileId;
        this.id = id;
        this.category = category;
        this.color = color;
        this.description = description;
        this.name = name;
        this.value = value;
        this.weight = weight;
        this.semester = semester;
        this.teacherId = teacherId;
        this.subjectId = subjectId;
    }

    /*@Ignore
    public Grade(int profileId, long id, String description, String name, float value, float weight, int semester, long teacherId, long categoryId, long subjectId) {
        this.profileId = profileId;
        this.id = id;
        this.description = description;
        this.name = name;
        this.value = value;
        this.weight = weight;
        this.semester = semester;
        this.teacherId = teacherId;
        //this.categoryId = categoryId;
        this.subjectId = subjectId;
    }*/
}

File: app/src/main/java/pl/szczodrzynski/edziennik/data/db/entity/Grade.kt
/*
 * Copyright (c) Kuba Szczodrzyński 2020-4-24.
 */
package pl.szczodrzynski.edziennik.data.db.entity

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.Index

@Entity(tableName = "grades",
        primaryKeys = ["profileId", "gradeId"],
        indices = [
            Index(value = ["profileId"])
        ])
open class Grade(
        val profileId: Int,
        @ColumnInfo(name = "gradeId")
        val id: Long,
        @ColumnInfo(name = "gradeName")
        var name: String,
        @ColumnInfo(name = "gradeType")
        var type: Int,

        @ColumnInfo(name = "gradeValue")
        var value: Float,
        @ColumnInfo(name = "gradeWeight")
        var weight: Float,
        @ColumnInfo(name = "gradeColor")
        var color: Int,

        @ColumnInfo(name = "gradeCategory")
        var category: String?,
        @ColumnInfo(name = "gradeDescription")
        var description: String?,
        @ColumnInfo(name = "gradeComment")
        var comment: String?,

        @ColumnInfo(name = "gradeSemester")
        val semester: Int,
        val teacherId: Long,
        val subjectId: Long,
        var addedDate: Long = System.currentTimeMillis()
) : Keepable(), Noteable {
    companion object {
        const val TYPE_NORMAL = 0
        const val TYPE_SEMESTER1_PROPOSED = 1
        const val TYPE_SEMESTER1_FINAL = 2
        const val TYPE_SEMESTER2_PROPOSED = 3
        const val TYPE_SEMESTER2_FINAL = 4
        const val TYPE_YEAR_PROPOSED = 5
        const val TYPE_YEAR_FINAL = 6
        const val TYPE_POINT_AVG = 10
        const val TYPE_POINT_SUM = 20
        const val TYPE_DESCRIPTIVE = 30
        const val TYPE_DESCRIPTIVE_TEXT = 31
        const val TYPE_TEXT = 40
    }

    @ColumnInfo(name = "gradeValueMax")
    var valueMax: Float? = null
    @ColumnInfo(name = "gradeClassAverage")
    var classAverage: Float? = null

    /**
     * Applies for historical grades. It's the new/replacement grade's ID.
     */
    @ColumnInfo(name = "gradeParentId")
    var parentId: Long? = null
    /**
     * Applies for current grades. If the grade was worse and this is the improved one.
     */
    @ColumnInfo(name = "gradeIsImprovement")
    var isImprovement = false

    @Ignore
    var showAsUnseen = false

    val isImproved
        get() = parentId ?: -1L != -1L

    override fun getNoteType() = Note.OwnerType.GRADE
}

